### PR TITLE
Implement feedreader service

### PR DESCRIPTION
**Pull request by Kegsay**
Notable improvements on this would be:
 - We currently do 1 goroutine per service. This could be bad if we have lots of these things running around.
 - We do not cache the response to RSS feeds. If we have 10 independent services on the same feed URL, we will
   hit the URL 10 times. This is similar to how we currently do 1 webhook/service, so it's plausible that in
   the future we will want to have some kind of generic caching layer.
 - The polling interval is divorced from the actual feed repoll time. Ideally we would schedule the goroutine
   only when we need it, rather than checking frequently, determining we have nothing to do, and going back
   to sleep.
 - The `Clients` instance is hackily added via a `SetClients` global. Given we *know* which client the service always wants, I wonder if we should be injecting this into the Service somehow.

### CODE REVIEW

File: src/github.com/matrix-org/go-neb/services/feedreader/feedreader.go [hunk 1]
```diff
@@ -0,0 +1,235 @@
...
+func (p *feedPoller) IntervalSecs() int64 { return 10 }
+func (p *feedPoller) OnPoll(s types.Service, cli *matrix.Client) {
+	logger := log.WithFields(log.Fields{
+		"service_id":   s.ServiceID(),
+		"service_type": s.ServiceType(),
+	})
+
+	frService, ok := s.(*feedReaderService)
+	if !ok {
+		logger.Error("FeedReader: OnPoll called without an Feed Service instance")
```
Comment (lukebarnard1) [inline src/github.com/matrix-org/go-neb/services/feedreader/feedreader.go]: s/an/a

File: src/github.com/matrix-org/go-neb/services/feedreader/feedreader.go [hunk 2]
```diff
@@ -0,0 +1,235 @@
...
+		return
+	}
+	now := time.Now().Unix() // Second resolution
+
+	// Work out which feeds should be polled
+	var pollFeeds []string
+	for u, feedInfo := range frService.Feeds {
```
Comment (lukebarnard1) [inline src/github.com/matrix-org/go-neb/services/feedreader/feedreader.go]: It took me awhile to figure out that ```u``` is a URL, despite that probably being a Go convention(?)

I also didn't realise ```pollFeeds``` was an array of URLs for awhile.

File: src/github.com/matrix-org/go-neb/services/feedreader/feedreader.go [hunk 3]
```diff
@@ -0,0 +1,235 @@
...
+					"feed_url":   u,
+					log.ErrorKey: err,
+					"item":       item,
+				}).Error("Failed to send item to room")
+			}
+		}
+	}
+
+	// Persist the service to save the next poll times if we did some queries
+	if len(pollFeeds) == 0 {
```
Comment (lukebarnard1) [inline src/github.com/matrix-org/go-neb/services/feedreader/feedreader.go]: I feel like this early return should go before the range over ```pollFields```

File: src/github.com/matrix-org/go-neb/services/feedreader/feedreader.go [hunk 4]
```diff
@@ -0,0 +1,235 @@
...
+			logger.WithError(err).WithField("room_id", roomID).Error("Failed to send to room")
+		}
+	}
+	return nil
+}
+
+// SomeOne posted a new article: Title Of The Entry ( https://someurl.com/blag )
+func itemToHTML(feed *gofeed.Feed, item gofeed.Item) matrix.HTMLMessage {
+	return matrix.GetHTMLMessage("m.notice", fmt.Sprintf(
+		"<i>%s</i> posted a new article: %s ( %s )",
```
Comment (lukebarnard1) [inline src/github.com/matrix-org/go-neb/services/feedreader/feedreader.go]: Will it always be "someone posting a new article"? If anything, we can only assume that a feed has been updated. I feel like this should be configurable somewhere...

File: src/github.com/matrix-org/go-neb/services/feedreader/feedreader.go [hunk 5]
```diff
@@ -0,0 +1,235 @@
...
+func (s *feedReaderService) ServiceID() string     { return s.id }
+func (s *feedReaderService) ServiceType() string   { return "feedreader" }
+func (s *feedReaderService) Poller() types.Poller  { return &feedPoller{} }
+
+// Register will check the liveness of each RSS feed given. If all feeds check out okay, no error is returned.
+func (s *feedReaderService) Register(oldService types.Service, client *matrix.Client) error {
+	if len(s.Feeds) == 0 {
+		// this is an error UNLESS the old service had some feeds in which case they are deleting us :(
+		var numOldFeeds int
+		oldFeedService, ok := oldService.(*feedReaderService)
```
Comment (lukebarnard1) [inline src/github.com/matrix-org/go-neb/services/feedreader/feedreader.go]: Not sure what this syntax is...

File: src/github.com/matrix-org/go-neb/services/feedreader/feedreader.go [hunk 6]
```diff
@@ -0,0 +1,235 @@
...
+	return nil
+}
+
+func (s *feedReaderService) PostRegister(oldService types.Service) {
+	if len(s.Feeds) == 0 { // bye-bye :(
+		logger := log.WithFields(log.Fields{
+			"service_id":   s.ServiceID(),
+			"service_type": s.ServiceType(),
+		})
+		logger.Info("Deleting service (0 feeds)")
```
Comment (lukebarnard1) [inline src/github.com/matrix-org/go-neb/services/feedreader/feedreader.go]: ```0 feeds``` might look like 0 is a formatted value. What about, "no feeds remaining"?

File: src/github.com/matrix-org/go-neb/services/feedreader/feedreader.go [hunk 7]
```diff
@@ -0,0 +1,235 @@
...
+		polling.StopPolling(s)
+		if err := database.GetServiceDB().DeleteService(s.ServiceID()); err != nil {
+			logger.WithError(err).Error("Failed to delete service")
+		}
+	}
+}
+
+func init() {
+	types.RegisterService(func(serviceID, serviceUserID, webhookEndpointURL string) types.Service {
```
Comment (lukebarnard1) [inline src/github.com/matrix-org/go-neb/services/feedreader/feedreader.go]: Not sure what ```webhookEndpointURL```  is

File: src/github.com/matrix-org/go-neb/services/feedreader/feedreader.go (hunk 2 above)
Comment (Kegsay) [inline src/github.com/matrix-org/go-neb/services/feedreader/feedreader.go]: It's a Go convention to be terse with variable names (they prefer single letters or just a few).

File: src/github.com/matrix-org/go-neb/services/feedreader/feedreader.go (hunk 3 above)
Comment (Kegsay) [inline src/github.com/matrix-org/go-neb/services/feedreader/feedreader.go]: Agreed.

File: src/github.com/matrix-org/go-neb/services/feedreader/feedreader.go (hunk 4 above)
Comment (Kegsay) [inline src/github.com/matrix-org/go-neb/services/feedreader/feedreader.go]: We don't currently support templates in Go-NEB. I think we will want to in the future, so people can configure their GH messages, RSS messages, JIRA messages, etc. But that doesn't exist yet.

File: src/github.com/matrix-org/go-neb/services/feedreader/feedreader.go (hunk 7 above)
Comment (Kegsay) [inline src/github.com/matrix-org/go-neb/services/feedreader/feedreader.go]: Not relevant to this particular service, but that is the function signature of the factory method.

File: src/github.com/matrix-org/go-neb/services/feedreader/feedreader.go (hunk 1 above)
Comment (Kegsay) [inline src/github.com/matrix-org/go-neb/services/feedreader/feedreader.go]: Done.

File: src/github.com/matrix-org/go-neb/services/feedreader/feedreader.go (hunk 3 above)
Comment (Kegsay) [inline src/github.com/matrix-org/go-neb/services/feedreader/feedreader.go]: Done.

File: src/github.com/matrix-org/go-neb/services/feedreader/feedreader.go (hunk 5 above)
Comment (lukebarnard1) [inline src/github.com/matrix-org/go-neb/services/feedreader/feedreader.go]: Ah, ok.

**Comment by Kegsay** [comment]
Anything left to do here?

**Comment by Kegsay** [comment]
IRL'd, nope!

**Pull request merged**